holmes/core/truncation/compaction.py

```
@@ -127,47 +249,82 @@ def compact_conversation_history(
+        # Compatibility fallback: some gateways mis-translate tool blocks / tools
+        # (ROB-424 was reported on Kong AI Gateway and vanilla LiteLLM proxies).
+        # Flattening tool messages to text and sending no tools is accepted by
+        # every OpenAI-compatible endpoint.
+        logging.warning(
+            f"Compaction: primary summarization attempt unusable ({fallback_reason}); "
+            "retrying with tool messages flattened to text and no tools attached"
+        )
+        flattened_history, _ = strip_system_prompt(conversation_history)
+        flattened_history = _flatten_tool_messages_for_compaction(flattened_history)
+        response = llm.completion(
+            messages=flattened_history + [instructions_message], drop_params=True
```

Ruff (0.15.21)

[warning] Consider `[*flattened_history, instructions_message]` instead of concatenation

Replace with `[*flattened_history, instructions_message]`

(RUF005)